pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -210,26 +248,50 @@ func (t *ResolvedPipelineRunTask) conditionsSkip() bool {
 	return false
 }
 
-func (t *ResolvedPipelineRunTask) whenExpressionsSkip(facts *PipelineRunFacts) bool {
+// skipBecauseWhenExpressionsEvaluatedToFalse confirms that the when expressions have completed evaluating, and
+// it returns true if any of the when expressions evaluate to false
+func (t *ResolvedPipelineRunTask) skipBecauseWhenExpressionsEvaluatedToFalse(facts *PipelineRunFacts) bool {
 	if t.checkParentsDone(facts) {
-		if len(t.PipelineTask.WhenExpressions) > 0 {
-			if !t.PipelineTask.WhenExpressions.HaveVariables() {
-				if !t.PipelineTask.WhenExpressions.AllowsExecution() {
-					return true
-				}
-			}
+		if !t.PipelineTask.WhenExpressions.AllowsExecution() {
+			return true
 		}
 	}
 	return false
 }
 
-func (t *ResolvedPipelineRunTask) parentTasksSkip(facts *PipelineRunFacts) bool {
+// skipBecauseParentTaskWasSkipped loops through the parent tasks and checks if the parent task skipped:
+//    if yes, is it because of when expressions and are when expressions?
+//        if yes, it ignores this parent skip and continue evaluating other parent tasks
+//        if no, it returns true to skip the current task because this parent task was skipped
+//    if no, it continues checking the other parent tasks
+func (t *ResolvedPipelineRunTask) skipBecauseParentTaskWasSkipped(facts *PipelineRunFacts) bool {
 	stateMap := facts.State.ToMap()
 	node := facts.TasksGraph.Nodes[t.PipelineTask.Name]
 	for _, p := range node.Prev {
-		if stateMap[p.Task.HashKey()].Skip(facts) {
+		parentTask := stateMap[p.Task.HashKey()]
+		if parentSkipStatus := parentTask.Skip(facts); parentSkipStatus.IsSkipped {
+			// if the `when` expressions are scoped to task and the parent task was skipped due to its `when` expressions,
+			// then we should ignore that and continue evaluating if we should skip because of other parent tasks
+			if parentSkipStatus.SkippingReason == WhenExpressionsSkip && facts.ScopeWhenExpressionsToTask {
+				continue
+			}
+			return true
+		}
+	}
+	return false
+}
+
+// skipBecauseResultReferencesAreMissing checks if the task references results that cannot be resolved, which is a
+// reason for skipping the task, and applies result references if found
+func (t *ResolvedPipelineRunTask) skipBecauseResultReferencesAreMissing(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) && t.hasResultReferences() {
+		resolvedResultRefs, pt, err := ResolveResultRefs(facts.State, PipelineRunState{t})
+		rprt := facts.State.ToMap()[pt]
+		if err != nil && (t.IsFinalTask(facts) || rprt.Skip(facts).SkippingReason == WhenExpressionsSkip) {
```

Comment:
Hey @jerop i don’t think we need to limit this check to finally 🤔, in case of the DAG tasks, a task is skipped without any when expression or without having when expression in any of the parents.

the other alpha test is failing with:

‘’’
examples_test.go:65: Failed waiting for task run done: timed out waiting for the condition
‘’’

or is it a flake 🤔

Reply:
that was a flake, i tested it locally and it passed

Reply:
it just passed here too: https://tekton-releases.appspot.com/build/tekton-prow/pr-logs/pull/tektoncd_pipeline/4085/pull-tekton-pipeline-alpha-integration-tests/1425449864539410432/

Reply:
it's not limited to finally only, we check whether:
- current task is a finally task has missing result references
- current task has missing result references from a parent task that was skipped due to its whenexpressions

Reply:
Awesome 👏 thanks

Reply:
/hold cancel